ksql-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -76,47 +81,62 @@ public PlanNode buildPlan() {
       currentNode = buildProjectNode(currentNode.getSchema(), currentNode);
     }
 
-    return buildOutputNode(currentNode.getSchema(), currentNode);
+    return buildOutputNode(
+        currentNode.getSchema(),
+        currentNode);
   }
 
-  private OutputNode buildOutputNode(final Schema inputSchema, final PlanNode sourcePlanNode) {
+  private OutputNode buildOutputNode(final Schema inputSchema,
+                                     final PlanNode sourcePlanNode) {
     StructuredDataSource intoDataSource = analysis.getInto();
 
+    final Map<String, Object> intoProperties = analysis.getIntoProperties();
+    final TimestampExtractionPolicy extractionPolicy = getTimestampExtractionPolicy(
+        sourcePlanNode,
+        inputSchema,
+        intoProperties);
     if (intoDataSource instanceof KsqlStdOut) {
       return new KsqlBareOutputNode(
           new PlanNodeId(KsqlStdOut.KSQL_STDOUT_NAME),
           sourcePlanNode,
           inputSchema,
-          analysis.getLimitClause()
+          analysis.getLimitClause(),
+          extractionPolicy
       );
     } else if (intoDataSource != null) {
-      Field timestampField = null;
-      if (analysis.getIntoProperties().get(KsqlConstants.SINK_TIMESTAMP_COLUMN_NAME) != null) {
-        timestampField =
-            SchemaUtil.getFieldByName(
-                inputSchema,
-                analysis.getIntoProperties()
-                    .get(KsqlConstants.SINK_TIMESTAMP_COLUMN_NAME)
-                    .toString()
-            ).get();
-      }
-
       return new KsqlStructuredDataOutputNode(
           new PlanNodeId(intoDataSource.getName()),
           sourcePlanNode,
           inputSchema,
-          timestampField,
+          extractionPolicy,
           sourcePlanNode.getKeyField(),
           intoDataSource.getKsqlTopic(),
           intoDataSource.getKsqlTopic().getTopicName(),
-          analysis.getIntoProperties(),
+          intoProperties,
           analysis.getLimitClause()
       );
 
     }
     throw new RuntimeException("INTO clause is not supported in SELECT.");
   }
 
+  private TimestampExtractionPolicy getTimestampExtractionPolicy(
+      final PlanNode sourcePlanNode,
+      final Schema inputSchema,
+      final Map<String, Object> intoProperties) {
+
+    final TimestampExtractionPolicy extractionPolicy
+        = TimestampExtractionPolicyFactory.create(inputSchema,
+        (String) intoProperties.get(KsqlConstants.SINK_TIMESTAMP_COLUMN_NAME),
+        (String) intoProperties.get(DdlConfig.TIMESTAMP_FORMAT_PROPERTY));
+
+    if (extractionPolicy instanceof MetadataTimestampExtractionPolicy) {
+      return sourcePlanNode.getTheSourceNode().getTimestampExtractionPolicy();
```

Comment:
I don't think it makes sense to inherit the source's extraction policy. What if the created stream doesn't have the timestamp field or has a field with the same name as the source's timestamp field but with data that's not really a timestamp? Streams produces results to the sink topic with the extracted timestamp as the message timestamp, so why not just use that if the user hasn't specified anything?

Reply:
i did this so long ago that i have to try and remember the reasoning behind it!

Reply:
I *think* the reasoning was this:
if the source node is not a CTAS/CSAS, i.e., it is just a CREATE TABLE or CREATE STREAM then we have to use the extraction policy as defined by that node. As the source topic is the one defined by the DDL command and not one that KSQL has written to.
For CTAS/CSAS - i'm not sure. I'll have to look into it